akka-stream/src/main/scala/akka/stream/impl/io/OutputStreamSourceStage.scala

```diff
@@ -41,159 +41,59 @@ final private[stream] class OutputStreamSourceStage(writeTimeout: FiniteDuration
-          Future {
-            val currentThread = Thread.currentThread()
-            // keep track of the thread for postStop interrupt
-            blockingThreadRef.compareAndSet(null, currentThread)
-            try {
-              dataQueue.take()
-            } catch {
-              case _: InterruptedException ⇒
-                Thread.interrupted()
-                ByteString.empty
-            } finally {
-              blockingThreadRef.compareAndSet(currentThread, null);
-            }
-          }.onComplete(downstreamCallback.invoke)
+          semaphore.release()
         }
       })
-
-      override def postStop(): Unit = {
-        //assuming there can be no further in messages
-        downstreamStatus.set(Canceled)
-        dataQueue.clear()
-        // if blocked reading, make sure the take() completes
-        dataQueue.put(ByteString.empty)
-        // interrupt any pending blocking take
-        val blockingThread = blockingThreadRef.get()
-        if (blockingThread != null)
-          blockingThread.interrupt()
-        super.postStop()
-      }
     }
 
     val logic = new OutputStreamSourceLogic
-    (logic, new OutputStreamAdapter(dataQueue, downstreamStatus, logic.wakeUp, writeTimeout))
+    (logic, new OutputStreamAdapter(semaphore, logic.upstreamCallback, writeTimeout))
   }
 }
 
 private[akka] class OutputStreamAdapter(
-  dataQueue:        BlockingQueue[ByteString],
-  downstreamStatus: AtomicReference[DownstreamStatus],
-  sendToStage:      (AdapterToStageMessage) ⇒ Future[Unit],
-  writeTimeout:     FiniteDuration)
+  unfulfilledDemand: Semaphore,
+  sendToStage:       AsyncCallback[AdapterToStageMessage],
+  writeTimeout:      FiniteDuration)
   extends OutputStream {
 
-  var isActive = true
-  var isPublisherAlive = true
-  def publisherClosedException = new IOException("Reactive stream is terminated, no writes are possible")
-
   @scala.throws(classOf[IOException])
-  private[this] def send(sendAction: () ⇒ Unit): Unit = {
-    if (isActive) {
-      if (isPublisherAlive) sendAction()
-      else throw publisherClosedException
-    } else throw new IOException("OutputStream is closed")
-  }
+  private[this] def sendData(data: ByteString): Unit = {
+    if (!unfulfilledDemand.tryAcquire(writeTimeout.toMillis, TimeUnit.MILLISECONDS)) {
+      throw new IOException("Timed out trying to write data to stream")
+    }
 
-  @scala.throws(classOf[IOException])
-  private[this] def sendData(data: ByteString): Unit =
-    send(() ⇒ {
-      try {
-        dataQueue.put(data)
-      } catch { case NonFatal(ex) ⇒ throw new IOException(ex) }
-      if (downstreamStatus.get() == Canceled) {
-        isPublisherAlive = false
-        throw publisherClosedException
-      }
-    })
+    val invocationResult =
+      sendToStage.invokeWithFeedback(Send(data))
+        .recoverWith {
+          case NonFatal(e) ⇒ throw new IOException(e)
+        }(ExecutionContexts.sameThreadExecutionContext)
```

Comment:
Same thread must not be used when an exception is thrown if I don't misremember

Reply:
But since we block on it below we can just try-catch there instead?

Reply:
Nicer still! :+1: